src/finch/finch_logic/nodes.py

```diff
@@ -451,53 +458,53 @@ class Produces(LogicNode):
         args: The arguments to return.
     """
 
-    args: tuple[LogicNode]
-
-    @staticmethod
-    def is_expr():
-        """Determines if the node is an expression."""
-        return True
+    args: tuple[LogicNode, ...]
 
     @staticmethod
     def is_stateful():
         """Determines if the node is stateful."""
         return True
 
-    def children(self):
+    def children(self) -> list[LogicNode]:  # type: ignore[override]
         """Returns the children of the node."""
         return [*self.args]
 
+    def get_fields(self) -> list[Field]:
+        raise NotImplementedError
+
     @classmethod
-    def make_term(cls, head, *args):
-        return head(args)
+    def make_term(  # type: ignore[override]
+        cls, head: Callable[[tuple[LogicNode, ...]], Self], *args: LogicNode
+    ) -> Self:
+        return head(tuple(args))
 
 
 @dataclass(eq=True, frozen=True)
-class Plan(LogicNode):
+class Plan(LogicExpression):
```

Comment:
There are two meanings of the word "expression" that we're working with here:
1. "S-Expression", a program node in an ast with a head and children.
2. "Expression" which is a pure computation that returns a value, as opposed to a "statement" which modifies program state.

I was assuming when you implemented get_fields for LogicExpression that you were referring to meaning 2. Now that i see a Plan is also a LogicExpression, I wonder if you should add "LogicStatement" as well, since Plan, Query, and Produces do not have fields, do not evaluate to tensors, and modify program state.